### ISSUE TITLE
Features request - Conformal Prediction

### ISSUE THREAD

**Issue by valeman** [catboost/catboost#2575]
Problem: CatBoost is a great library, but it currently lacks reliable modern uncertainty quantification that is rather easy to implement using conformal prediction. https://github.com/valeman/awesome-conformal-prediction

Feature request - add conformal prediction for regression and classification similar to what other libraries like MAPIE implemented.

1) Regression - Inductive (split) and Conformalized Quantile Regression https://mapie.readthedocs.io/en/stable/examples_regression/4-tutorials/plot_cqr_tutorial.html

2) Classification - Venn Abers

https://proceedings.neurips.cc/paper/2015/file/a9a1d5317a33ae8cef33961c34144f84-Paper.pdf

There is also talk by Vovk at Yandex https://m.youtube.com/watch?v=ksrUJdb2tA8&pp=ygUNVm92ayB2bGFkaW1pcg%3D%3D

Tutorial https://cml.rhul.ac.uk/copa2017/presentations/VennTutorialCOPA2017.pdf

https://github.com/ip200/venn-abers implementation by Ivan Petej 

Older implementation by Paolo Tocacceli

https://github.com/ptocca/VennABERS

Venn ABERS demo https://github.com/ptocca/VennABERS-demo

**Comment by andrey-khropov** [comment]
@valeman Are you familiar with already implemented methods to estimate uncertainty in CatBoost: https://catboost.ai/en/docs/references/uncertainty ?

**Comment by valeman** [comment]
@andrey-khropov yes - they seem to be ad hoc methods that do not have validity guarantees, here it the study about NG Boost 'Does NGBoost work? Evaluating NGBoost against key criteria for good probabilistic prediction' showing it produces biased predictions. 

https://medium.com/@valeman/does-ngboost-work-evaluating-ngboost-against-critical-criteria-for-good-probabilistic-prediction-28c4871c1bab

**Comment by aangelopoulos** [comment]
Hey @andrey-khropov! I'm Anastasios, a Ph.D. student at UC Berkeley.

I have been working on the conformal area for several years now, and would be happy to discuss the similarities/differences with the currently implemented methods on a technical level, if that would be helpful.

There would be a pretty big value-add to your user base, in my opinion. Conformal would be compatible with your existing UQ infra, and allow for formal "distribution-free" frequentist coverage guarantees---ones that don't assume the machine learning model is perfect/consistent, or anything about the particular parametric form of the data.

If you want to have an offline conversation, feel free to e-mail me at angelopoulos@berkeley.edu as well.

Cheers!

**Comment by valeman** [comment]
@aangelopoulos “as well as the exact variants that would be most useful to implement”. I guess you want to promote your methods developed for computer vision? Well CatBoost is tabular data library first and foremost and computer vision sets are unsuitable for tabular data.

I am pretty sure CatBoost team is fully capable implementing CQR and Venn Abers that are the best methods for tabular data, especially considering that other libraries cited in my list already implemented them.

**Comment by valeman** [comment]
There is a Kaggle competition in process where people tried to use CatBoost but the uncertainty estimation methods in CatBoost result in coverage much lower than that specified by the user https://www.kaggle.com/competitions/probabilistic-forecasting-i-temperature/discussion/493220

**Comment by FrancescoMandru** [comment]
Any update on this?

**Comment by kmedved** [comment]
This would be a nice additional, despite its somewhat redundant nature with RMSEWithUncertainty. MAPIE is a nice library for this, but struggles with large datasets, and Catboost has enough bespoke functionality that a built-in conformal prediction functionality would be helpful.

**Comment by GabrielFo** [comment]
It would indeed be a great feature for CatBoost, looking forward to it.

**Comment by jacktang** [comment]
hello, any progress on this?
